plugins/webp-uploads/picture-element.php

```diff
@@ -22,6 +22,15 @@ function webp_uploads_wrap_image_in_picture( string $image, string $context, int
 	if ( 'the_content' !== $context ) {
 		return $image;
 	}
+
+	// Bail early if picture element requirement is not met.
+	if ( ! webp_uploads_is_picture_element_enabled() ) {
+		return $image;
+	}
+
+	// Remove IMG replacement for picture element.
+	remove_filter( 'the_content', 'webp_uploads_update_image_references', 13 );
```

Comment:
This doesn't seem right. The filter is getting re-removed every time the `webp_uploads_wrap_image_in_picture()` function calls (which is every time `the_content` filters apply). You could modify `webp_uploads_update_image_references()` to also short-circuit if `webp_uploads_is_picture_element_enabled()` returns true. However, it would be more efficient if you only added the filters when we know they are needed. So for example, in `hooks.php`:

```php
add_action( 'init', static function () {
    if ( webp_uploads_is_picture_element_enabled() ) {
        add_filter( 'wp_content_img_tag', 'webp_uploads_wrap_image_in_picture', 10, 3 );
    } else {
        add_filter( 'the_content', 'webp_uploads_update_image_references', 13 ); // Run after wp_filter_content_tags.
    }
} );
```

(I'm using a closure as an example only. A named function would be used in commit.)

The above would eliminate this line:

https://github.com/WordPress/performance/blob/e34711dbf286a16523b3c3d35462e0e79ea4c6f9/plugins/webp-uploads/hooks.php#L579

And replace lines (modified above in this PR):

https://github.com/WordPress/performance/blob/e34711dbf286a16523b3c3d35462e0e79ea4c6f9/plugins/webp-uploads/hooks.php#L768-L770

Then there would be no need for this code here in your PR:

https://github.com/WordPress/performance/blob/37901ab13e8a8ed10c0d47647ca16a1b0da70229/plugins/webp-uploads/picture-element.php#L26-L32

Reply:
Thanks @westonruter If i add condition filter in `init` action the tests going failed and didn't return picture elements.

Init action:
```

```
1) Test_WebP_Uploads_Picture_Element::test_maybe_wrap_images_in_picture_element with data set "jpeg and picture enabled" (true, true, '<picture class="wp-picture-{{...cture>')
Failed asserting that two strings are equal.
--- Expected
+++ Actual
@@ @@
-'<picture class="wp-picture-4" style="display: contents;"><source type="image/webp" srcset="http://localhost:8889/wp-content/uploads/2024/07/leaves-1214-1024x683-jpg.webp 1024w, http://localhost:8889/wp-content/uploads/2024/07/leaves-1214-300x200-jpg.webp 300w, http://localhost:8889/wp-content/uploads/2024/07/leaves-1214-768x512-jpg.webp 768w" sizes="(max-width: 1024px) 100vw, 1024px"><source type="image/jpeg" srcset="http://localhost:8889/wp-content/uploads/2024/07/leaves-1214-1024x683.jpg 1024w, http://localhost:8889/wp-content/uploads/2024/07/leaves-1214-300x200.jpg 300w, http://localhost:8889/wp-content/uploads/2024/07/leaves-1214-768x512.jpg 768w" sizes="(max-width: 1024px) 100vw, 1024px"><img width="1024" height="683" src="http://localhost:8889/wp-content/uploads/2024/07/leaves-1214-1024x683.jpg" class="wp-image-4" alt="Green Leaves" decoding="async" loading="lazy" srcset="http://localhost:8889/wp-content/uploads/2024/07/leaves-1214-1024x683.jpg 1024w, http://localhost:8889/wp-content/uploads/2024/07/leaves-1214-300x200.jpg 300w, http://localhost:8889/wp-content/uploads/2024/07/leaves-1214-768x512.jpg 768w, http://localhost:8889/wp-content/uploads/2024/07/leaves-1214.jpg 1080w" sizes="(max-width: 1024px) 100vw, 1024px" /></picture>'
+'<img width="1024" height="683" src="http://localhost:8889/wp-content/uploads/2024/07/leaves-1214-1024x683.jpg" class="wp-image-4" alt="Green Leaves" decoding="async" loading="lazy" srcset="http://localhost:8889/wp-content/uploads/2024/07/leaves-1214-1024x683.jpg 1024w, http://localhost:8889/wp-content/uploads/2024/07/leaves-1214-300x200.jpg 300w, http://localhost:8889/wp-content/uploads/2024/07/leaves-1214-768x512.jpg 768w, http://localhost:8889/wp-content/uploads/2024/07/leaves-1214.jpg 1080w" sizes="(max-width: 1024px) 100vw, 1024px" />'

/var/www/html/wp-content/plugins/performance/plugins/webp-uploads/tests/test-picture-element.php:92
```

Short-circuit in `webp_uploads_update_image_references` filter works fine.

Reply:
Yes, in your tests you'd need to remove the filters that `webp_uploads_init()` adds, and then you'd need to set the options to opt-in to the picture element, and then you'd re-call `webp_uploads_init()`. This logic would be put into the `opt_in_to_picture_element()` helper method.

Reply:
Addressed in https://github.com/WordPress/performance/pull/1408/commits/f93d0693fc257bd3b16393909e7b27d47d122644